test/hotspot/jtreg/runtime/os/TestTracePageSizes.java

```diff
@@ -93,32 +97,91 @@
+        return p2;
+    }
 
     // Parse /proc/self/smaps using a regexp capturing the address
     // ranges, what page size they have and if they might use
     // transparent huge pages. The pattern is not greedy and will
     // match as little as possible so each "segment" in the file
     // will generate a match.
     private static void parseSmaps() throws Exception {
-        String smapsPatternString = "(\\w+)-(\\w+).*?" +
-                                    "KernelPageSize:\\s*(\\d*) kB.*?" +
-                                    "VmFlags: ([\\w\\? ]*)";
-        Pattern smapsPattern = Pattern.compile(smapsPatternString, Pattern.DOTALL);
-        Scanner smapsScanner = new Scanner(new File("/proc/self/smaps"));
-        // Find all memory segments in the smaps-file.
-        smapsScanner.findAll(smapsPattern).forEach(mr -> {
-            String start = mr.group(1);
-            String end = mr.group(2);
-            String ps = mr.group(3);
-            String vmFlags = mr.group(4);
-
-            // Create a range given the match and add it to the list.
-            RangeWithPageSize range = new RangeWithPageSize(start, end, ps, vmFlags);
-            ranges.add(range);
-            debug("Added range: " + range);
-        });
-        smapsScanner.close();
+        // We can override the smaps file to parse to pass in a pre-fetched one
+        String smapsFileToParse = System.getProperty("smaps-file");
+        if (smapsFileToParse != null) {
+            parseSmaps(Paths.get(smapsFileToParse));
+        } else {
+            Path smapsCopy = copySmaps();
+            parseSmaps(smapsCopy);
+        }
+    }
+
+    static class SmapsParser {
+        static final Pattern SECTION_START_PATT = Pattern.compile("^([a-f0-9]+)-([a-f0-9]+) [\\-rwpsx]{4}.*");
+        static final Pattern KERNEL_PAGESIZE_PATT = Pattern.compile("^KernelPageSize:\\s*(\\d*) kB");
+        static final Pattern VMFLAGS_PATT = Pattern.compile("^VmFlags: ([\\w\\? ]*)");        String start;
+        String end;
+        String ps;
+        String vmFlags;
+        int lineno = 0;
+
+        void reset() {
+            start = null;
+            end = null;
+            ps = null;
+            vmFlags = null;
+        }
+
+        public void finish() {
+            if (start != null) {
+                RangeWithPageSize range = new RangeWithPageSize(start, end, ps, vmFlags);
+                ranges.add(range);
+                debug("Added range: " + range);
+                reset();
+            }
+        }
+
+        void eatNext(String line) {
+            debug("" + (lineno++) + " " + line);
+            Matcher matSectionStart = SECTION_START_PATT.matcher(line);
+            if (matSectionStart.matches()) {
+                finish();
+                start = matSectionStart.group(1);
+                end = matSectionStart.group(2);
+                ps = null;
+                vmFlags = null;
```

Comment:
I can see that `Range` code handles `vmFlags = null` specifically. Does it handle `ps == null`? I cannot see it.